File: core/src/main/java/io/snabble/sdk/checkout/CheckoutRetryer.java
package io.snabble.sdk.checkout;

import android.content.Context;
import android.content.SharedPreferences;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;

import com.google.gson.reflect.TypeToken;

import java.util.Date;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.CountDownLatch;

import io.snabble.sdk.PaymentMethod;
import io.snabble.sdk.Product;
import io.snabble.sdk.Project;
import io.snabble.sdk.ShoppingCart;
import io.snabble.sdk.Snabble;
import io.snabble.sdk.utils.Dispatch;
import io.snabble.sdk.utils.GsonHolder;
import io.snabble.sdk.utils.Logger;

class CheckoutRetryer {
    private class SavedCart {
        ShoppingCart.BackendCart backendCart;
        Date finalizedAt;
        int failureCount;

        SavedCart(ShoppingCart.BackendCart backendCart, Date finalizedAt) {
            this.backendCart = backendCart;
            this.finalizedAt = finalizedAt;
        }
    }

    private final PaymentMethod fallbackPaymentMethod;
    private final SharedPreferences sharedPreferences;
    private final Project project;
    private final CopyOnWriteArrayList<SavedCart> savedCarts;
    private CountDownLatch countDownLatch;

    CheckoutRetryer(Project project, PaymentMethod fallbackPaymentMethod) {
        Context context = Snabble.getInstance().getApplication();
        this.sharedPreferences = context.getSharedPreferences("snabble_saved_checkouts_" + project.getId(), Context.MODE_PRIVATE);
        this.project = project;
        this.fallbackPaymentMethod = fallbackPaymentMethod;

        String json = sharedPreferences.getString("saved_carts", null);
        if (json != null) {
            TypeToken typeToken = new TypeToken<CopyOnWriteArrayList<SavedCart>>() {};
            savedCarts = GsonHolder.get().fromJson(json, typeToken.getType());
        } else {
            savedCarts = new CopyOnWriteArrayList<>();
        }

        processPendingCheckouts();
    }

    public void add(final ShoppingCart.BackendCart backendCart) {
        Dispatch.mainThread(() -> {
            savedCarts.add(new SavedCart(backendCart, new Date()));
            save();
        });
    }

    private void save() {
        Dispatch.mainThread(() -> {
            String json = GsonHolder.get().toJson(savedCarts);

            sharedPreferences.edit()
                    .putString("saved_carts", json)
                    .apply();
        });
    }

    public void processPendingCheckouts() {
        Context context = Snabble.getInstance().getApplication();
        ConnectivityManager cm = (ConnectivityManager)context.getSystemService(Context.CONNECTIVITY_SERVICE);

        if (cm == null) {
            return;
        }

        NetworkInfo networkInfo = cm.getActiveNetworkInfo();
        if (networkInfo == null || !networkInfo.isConnected()) {
            return;
        }

        Dispatch.mainThread(() -> {
            if (countDownLatch != null && countDownLatch.getCount() > 0) {
                return;
            }

            countDownLatch = new CountDownLatch(savedCarts.size());

            for (final SavedCart savedCart : savedCarts) {
                if (savedCart.failureCount >= 3) {
                    removeSavedCart(savedCart);
                }

                final DefaultCheckoutApi checkoutApi = new DefaultCheckoutApi(project, project.getShoppingCart());
                checkoutApi.createCheckoutInfo(savedCart.backendCart, null, new CheckoutInfoResult() {
                    @Override
                    public void success(SignedCheckoutInfo signedCheckoutInfo, int onlinePrice, List<PaymentMethodInfo> availablePaymentMethods) {
                        checkoutApi.createPaymentProcess(UUID.randomUUID().toString(), signedCheckoutInfo, fallbackPaymentMethod, null,
                                true, savedCart.finalizedAt, new PaymentProcessResult() {
                            @Override
                            public void success(CheckoutProcessResponse checkoutProcessResponse, String rawResponse) {
                                Logger.d("Successfully resend checkout " + savedCart.backendCart.session);
                                removeSavedCart(savedCart);
                                countDownLatch.countDown();
                            }

                            @Override
                            public void error() {
                                fail();
                            }
                        });
                    }

                    @Override
                    public void noShop() {
                        fail();
                    }

                    @Override
                    public void invalidProducts(List<? extends Product> products) {
                        fail();
                    }

                    @Override
                    public void noAvailablePaymentMethod() {
                        fail();
                    }

                    @Override
                    public void invalidDepositReturnVoucher() {
                        fail();
                    }

                    @Override
                    public void unknownError() {
                        fail();
                    }

                    @Override
                    public void connectionError() {
                        fail();
                    }

                    private void fail() {
                        savedCart.failureCount++;
                        save();
                        countDownLatch.countDown();
                    }
                }, -1);
            }
        });
    }

    private void removeSavedCart(final SavedCart savedCart) {
        Dispatch.mainThread(() -> {
            savedCarts.remove(savedCart);
            save();
        });
    }
}

File: core/src/main/java/io/snabble/sdk/checkout/CheckoutRetryer.kt
package io.snabble.sdk.checkout

import android.content.Context
import android.content.SharedPreferences
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.Build
import com.google.gson.reflect.TypeToken
import io.snabble.sdk.PaymentMethod
import io.snabble.sdk.Product
import io.snabble.sdk.Project
import io.snabble.sdk.Snabble.instance
import io.snabble.sdk.shoppingcart.data.cart.BackendCart
import io.snabble.sdk.utils.Dispatch
import io.snabble.sdk.utils.GsonHolder
import io.snabble.sdk.utils.Logger
import java.util.*
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.CountDownLatch

internal class CheckoutRetryer(project: Project, fallbackPaymentMethod: PaymentMethod) {
    private inner class SavedCart(
        var backendCart: BackendCart,
        var finalizedAt: Date
    ) {

        var failureCount = 0
    }

    private val fallbackPaymentMethod: PaymentMethod
    private val sharedPreferences: SharedPreferences
    private val project: Project
    private var savedCarts = CopyOnWriteArrayList<SavedCart>()
    private var countDownLatch: CountDownLatch? = null

    init {
        val context: Context = instance.application
        sharedPreferences = context.getSharedPreferences(
            "snabble_saved_checkouts_" + project.id,
            Context.MODE_PRIVATE
        )

        this.project = project
        this.fallbackPaymentMethod = fallbackPaymentMethod
        val json = sharedPreferences.getString("saved_carts", null)

        savedCarts = if (json != null) {
            val typeToken: TypeToken<*> = object : TypeToken<CopyOnWriteArrayList<SavedCart?>?>() {}
            GsonHolder.get().fromJson(json, typeToken.type)
        } else {
            CopyOnWriteArrayList()
        }

        processPendingCheckouts()
    }

    fun add(backendCart: BackendCart) {
        Dispatch.mainThread {
            savedCarts.add(SavedCart(backendCart, Date()))
            save()
        }
    }

    private fun save() {
        Dispatch.mainThread {
            val json = GsonHolder.get().toJson(savedCarts)
            sharedPreferences.edit()
                .putString("saved_carts", json)
                .apply()
        }
    }

    private fun ConnectivityManager.isNetworkConnected(): Boolean {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            val networkCap = getNetworkCapabilities(activeNetwork) ?: return false
            listOf(
                NetworkCapabilities.TRANSPORT_WIFI,
                NetworkCapabilities.TRANSPORT_CELLULAR,
                NetworkCapabilities.TRANSPORT_ETHERNET,
                NetworkCapabilities.TRANSPORT_BLUETOOTH
            ).any(networkCap::hasTransport)
        } else {
            (@Suppress("DEPRECATION")
            activeNetworkInfo?.isConnected ?: false)
        }
    }

    fun processPendingCheckouts() {
        val context: Context = instance.application
        val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        if (cm.isNetworkConnected()) return

        Dispatch.mainThread {
            if ((countDownLatch?.count ?: 0) > 0) {
                return@mainThread
            }

            countDownLatch = CountDownLatch(savedCarts.size)
            savedCarts.forEach { savedCart ->
                if (savedCart.failureCount >= 3) {
                    removeSavedCart(savedCart)
                }

                val checkoutApi = DefaultCheckoutApi(project, project.shoppingCart)
                checkoutApi.createCheckoutInfo(savedCart.backendCart, object : CheckoutInfoResult {
                    override fun onSuccess(
                        signedCheckoutInfo: SignedCheckoutInfo,
                        onlinePrice: Int,
                        availablePaymentMethods: List<PaymentMethodInfo>
                    ) {
                        checkoutApi.createPaymentProcess(
                            id = UUID.randomUUID().toString(),
                            signedCheckoutInfo = signedCheckoutInfo,
                            paymentMethod = fallbackPaymentMethod,
                            processedOffline = true,
                            paymentCredentials = null,
                            finalizedAt = savedCart.finalizedAt,
                            paymentProcessResult = object : PaymentProcessResult {
                                override fun onSuccess(
                                    checkoutProcessResponse: CheckoutProcessResponse?,
                                    rawResponse: String?
                                ) {
                                    Logger.d("Successfully resend checkout " + savedCart.backendCart.session)
                                    removeSavedCart(savedCart)
                                    countDownLatch?.countDown()
                                }

                                override fun onError() {
                                    fail()
                                }

                                override fun onNotFound() {
                                    fail()
                                }
                            })
                    }

                    override fun onNoShopFound() {
                        fail()
                    }

                    override fun onInvalidProducts(products: List<Product>) {
                        fail()
                    }

                    override fun onNoAvailablePaymentMethodFound() {
                        fail()
                    }

                    override fun onUnknownError() {
                        fail()
                    }

                    override fun onConnectionError() {
                        fail()
                    }

                    private fun fail() {
                        savedCart.failureCount++
                        save()
                        countDownLatch?.countDown()
                    }
                }, -1)
            }
        }
    }

    private fun removeSavedCart(savedCart: SavedCart) {
        Dispatch.mainThread {
            savedCarts.remove(savedCart)
            save()
        }
    }
}
